src/openrct2/scripting/ScSocketServer.hpp

```diff
@@ -0,0 +1,505 @@
+                        _eventList.Raise(EVENT_CONNECT_ONCE, GetPlugin(), {}, false);
+                        _eventList.RemoveAllListeners(EVENT_CONNECT_ONCE);
+                    }
+                    else if (status == SOCKET_STATUS_CLOSED)
+                    {
+                        _connecting = false;
+                        _eventList.Raise(EVENT_ERROR, GetPlugin(), {}, false);
+                    }
+                }
+                else if (status == SOCKET_STATUS_CONNECTED)
+                {
+                    char buffer[2048];
+                    size_t bytesRead{};
+                    auto result = _socket->ReceiveData(buffer, sizeof(buffer), &bytesRead);
+                    switch (result)
+                    {
+                        case NETWORK_READPACKET_SUCCESS:
+                            RaiseOnData(std::string(buffer, bytesRead));
+                            break;
+                        case NETWORK_READPACKET_NO_DATA:
+                            break;
+                        case NETWORK_READPACKET_MORE_DATA:
+                            break;
+                        case NETWORK_READPACKET_DISCONNECTED:
+                            CloseSocket();
+                            _disposed = true;
+                            break;
+                    }
+                }
+                else
+                {
+                    CloseSocket();
+                    _disposed = true;
+                }
+            }
+        }
+
+        void Dispose() override
+        {
+            CloseSocket();
+            _disposed = true;
+        }
+
+        bool IsDisposed() const override
+        {
+            return _disposed;
+        }
+
+        static void Register(duk_context* ctx)
+        {
+            dukglue_register_method(ctx, &ScSocket::destroy, "destroy");
+            dukglue_register_method(ctx, &ScSocket::setNoDelay, "setNoDelay");
+            dukglue_register_method(ctx, &ScSocket::connect, "connect");
+            dukglue_register_method(ctx, &ScSocket::end, "end");
+            dukglue_register_method(ctx, &ScSocket::write, "write");
+            dukglue_register_method(ctx, &ScSocket::on, "on");
+            dukglue_register_method(ctx, &ScSocket::off, "off");
+        }
+    };
+
+    class ScSocketServer : public ScSocketBase
+    {
+    private:
+        static constexpr uint32_t EVENT_NONE = std::numeric_limits<uint32_t>::max();
+        static constexpr uint32_t EVENT_CONNECTION = 0;
+
+        EventList _eventList;
+        std::unique_ptr<ITcpSocket> _socket;
+        std::vector<std::shared_ptr<ScSocket>> _scClientSockets;
+        bool _disposed{};
+
+        bool listening_get()
+        {
+            if (_socket != nullptr)
+            {
+                return _socket->GetStatus() == SOCKET_STATUS_LISTENING;
+            }
+            return false;
+        }
+
+        ScSocketServer* close()
+        {
+            Dispose();
```

Comment:
There is the ambiguity of close vs dispose here again, closing a socket should not destroy it as one might use it again.